custom/icds_reports/tests/test_aggregation_script.py

```diff
@@ -0,0 +1,212 @@
+        if match:
+            return '0.{}'.format(int(match.group('zeros')) * '0')
+        else:
+            return value_str
+
+    def _load_data_from_db(self, table_name, sort_key, filter_by=None):
+        session_helper = connection_manager.get_session_helper(ICDS_UCR_CITUS_ENGINE_ID)
+        engine = session_helper.engine
+        session = session_helper.Session
+        metadata = sqlalchemy.MetaData(bind=engine)
+        metadata.reflect(bind=engine)
+        table = metadata.tables[table_name]
+        columns = [
+            column.name
+            for column in table.columns
+        ]
+        with engine.begin() as connection:
+            query = session.query(table).order_by(*sort_key)
+            if filter_by:
+                query = query.filter_by(**filter_by)
+            rows = query.all()
+            for row in rows:
+                row = list(row)
+                for idx, value in enumerate(row):
+                    if isinstance(value, date):
+                        row[idx] = value.strftime('%Y-%m-%d')
+                    elif isinstance(value, time):
+                        row[idx] = value.strftime("%H:%M:%S.%f").rstrip('0').rstrip('.')
+                    elif isinstance(value, int):
+                        row[idx] = str(value)
+                    elif isinstance(value, (float, Decimal)):
+                        row[idx] = self._convert_decimal_to_string(row[idx])
+                    elif value is None:
+                        row[idx] = ''
+                yield dict(zip(columns, row))
+
+    def _load_and_compare_data(self, table_name, path, sort_key=None, filter_by=None):
+        # To speed up tests, we use a sort_key wherever possible
+        #   to presort before comparing data
+        if sort_key:
+            self._fasterAssertListEqual(
+                list(self._load_data_from_db(table_name, sort_key, filter_by)),
+                self._load_csv(path)
+            )
+        else:
+            sort_key = lambda x: x
+            self._fasterAssertListEqual(
+                sorted(
+                    list(self._load_data_from_db(table_name, [], filter_by)),
+                    key=sort_key
+                ),
+                sorted(
+                    self._load_csv(path),
+                    key=sort_key
+                )
+            )
+
+
+class CcsRecordMonthlyAggregationTest(AggregationScriptTestBase):
+    always_include_columns = {'awc_id', 'case_id'}
+    fixtures_needed = [
+        'ccs_cases.csv',
+        'pregnant_tasks.csv',
+        'person_cases.csv',
+        'thr_form.csv',
+        'birth_preparedness.csv',
+        'pnc_forms.csv',
+        'complementary_feeding.csv',
+        'delivery_form.csv',
+    ]
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        for state_id in ('st1', 'st2'):
+            _aggregate_bp_forms(state_id, datetime(2017, 3, 31))
+
+        for month in _get_monthly_dates(date(2017, 5, 28), 2):
```

Comment:
Hey, Isolating the aggregation test cases to run partial parts of agg. Wont it increase the overall completion time of test cases. Because for example here it looks in setup it creates just ccs record tables and in teardown it destroys them. If I would want to run all agg test cases. It would have to rebuild all ccs record tables again. Same will happen when isolating child health test cases.
Does it make sense? I may be missing something obvious

Reply:
Yes, there are tradeoffs of what's going on here. Some factors I can think of:

* Full end-to-end integration test
* Isolation of tests so that each test tests one thing
* Ease of updating single tests
* Speed of running single tests

I think that by extracting the tests separately improves our capability of doing the bottom three. I would argue that we do not have a full end to end integration test (and I think from Cory's comment he agrees), because the way we currently do aggregation is not the same as airflow. We currently run the old celery task that is no longer used except for in tests. I think this is very bad and means that our workflow is easily broken. Ideally we should have an end to end test that uses airflow to actually run the aggregation instead of relying on an unused celery task.

Another factor here is the difficulty of updating current tests. Because the CSV files affect every test here, we often reason about the easiest way to update the CSVs instead of how to update them in a way that makes it easy to avoid updating every test (or at least that's how I've done it previously).

> Making it separate for separate usecases(mother and child) I think this will inherently make the test case updation easier But will make new changes less reliable and more prone to push wrong change without proper testing

I don't agree with this at the moment. As I said above, I have sometimes thought about how to update tests to optimize the amount of work I need to put in to updating the tests. How many reporting bug fixes have not had test updates, because it would be too difficult? Do you think that we could add more comprehensive tests, if that data was local to each test that failed?